pypesto/objective/amici.py

```diff
@@ -222,7 +222,8 @@ def __getstate__(self) -> Dict:
         # read in byte stream
         amici_solver_settings = open(amici_solver_file, 'rb').read()
         state['amici_solver_settings'] = amici_solver_settings
-        # remove temporary file
+        # close file descriptor and remove temporary file
+        os.close(fd)
```

Comment:
What about letting [tempfile.TemporaryFile](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryFile) handle that?

Reply:
It is a possibility (`NamedTemporaryFile` should be used since AMICI needs to access it too), doing so will open the file for I/O so that a `flush()` will probably be needed before reading from AMICI. I would personally prefer to keep the two things explicitly separate, but it is just a matter of taste and both solutions would be perfectly equivalent.

Reply:
The problem with the current implementation is that it's not exception safe. If anything `raise`s before `close`, nothing will be closed.

Reply:
Added a `try .. finally`.